### ISSUE TITLE
sepearation of authority + freedom to migrate

### ISSUE THREAD

**Issue by serapath** [pfrazee/ctzn#2]
> * As records are linked by their pubkey-URL, there is no hard binding to the example server.
>
> Because of this separation of authority, the members of example.com maintain the freedom to migrate their content away from the example community without any disruption to the network. This migration would be similar to "forking" a codebase -- the content would remain but the community authority would be modified.
>
> This interplay between community authority and individual authority is a repeatable PX pattern. By maintaining individual rights to migrate away from a community, the community maintainers' authority is contingent on the members' satisfaction with their leadership.

In blockchain terms I read it as a single proof-of-authority miner (a.k.a block producer), where all users run either light nodes (=not audit the blockchain and only sparsely downloading the content they need to use the app) or full nodes (=download and replay the entire blockchain to verify).

If a community smart contract violation has been detected - and the community is large (e.g. doesn't have to be twitter large) - then users can "take it or leave it" (leaving = loosing all their friends and connections in the network), but there seems to be no straight forward "built in" way to communicate with other users to coordinate for what the fork should be.

Which to me sounds not great if it stayed like that. It would make the "divide and conquer" principles apply or call it the "network effect" or whatever. 

It just means, migrating away from e.g. twitter or facebook or any other such app to a diffrent (maybe objectively better one) is impossible, because you can't take your friends/followers/follows with you. to the new place for lack of ability to coordinate on what that new app should be.

**Comment by pfrazee** [comment]
> because you can't take your friends/followers/follows with you

So this isn't accurate -- anything that has to do with personal data like your friends or followers will migrate with you wherever you go.

**Comment by serapath** [comment]
I didn't mean the data. You can copy the data, but what about the network?

You can fork a source code easily, but not a network.
You can copy the state of the network and run a parallel one (e.g. fork bitcoin into bitcoin2 with the same state)
But will the people now run nodes and connect and progress the state on bitcoin2 or will they stick to bitcoin?

**Comment by serapath** [comment]
It is a coordination problem, or not?

If one network would need to trust one or a few authorities and some would detect fraud and fork the network and tell everyone to move and show the proof of fraud, most people might not notice that news in all the daily noise and even if - the authorities might just spawn a gazillion more parallel networks and run a campaign to tlel people a different message about the "new alternative network".

...everyone would be utterly confused - which of the 1000 new networks is the new honest network?
...if i join any of those and people don't move with me, i might not be able to receive my usual feed of updates anymore and they might not receive my updates too, so i rather stick with the current network, because of all the people and the feed i get here and dont change despite the proof of fraud?

**Comment by pfrazee** [comment]
> But will the people now switch and run bitcoin2 nodes and progress the state on bitcoin2 or will they stay with bitcoin?

It's exactly the same as forking code!

If somebody forks `pfrazee/ctzn` to `alice/ctzn` then the question is whether people change their git remotes to `alice/ctzn`.  Likewise with a community, if there's a community at `server.com` and somebody forks it to `fork.com`, the server db would change but not the users' dbs, so then it's a question of whether the users move to `fork.com` when reading the community content. So it's very similar to code forking.

To give a more immediate example, if `r/WallStreetBets` users decided to fork to `bettit.com` it would happen by copying the community db. The bettit.com WSB would have all the same users' content as when the fork occurred, and continue to get content from the reddit version, but new moderation and etc would apply to what the bettit.com shows.

This is separate from having something like the ability to move a forum/community between servers, which is also possible in ctzn but suggests the admins of WSB configuring the move rather than the users forking WSB.

> ...everyone would be utterly confused - which of the 1000 new networks is the new honest network?

The confusing thing is understanding what communities are in CTZN. Communities are basically indexes of user content. The people who admin a community are making moderation and ranking decisions which affect those indexes. The users' content -- the primary records -- all live separately in their own DBs.

So when you're evaluating which is the "real" community, you'll be looking at the DNS-IDs. Is it `WSB@reddit.com` or `WSB@bettit.com`?

**Comment by serapath** [comment]
I agree with everything you describe. But I still see a fundamental qualitative difference.

If somebody forks a repo to use a modified version of it, it is immediately useful.
Whether it is a module dependency or an app like google docs or vscode, a fork - given you can maintain it, makes sense.
It doesn't really matter if other users agree with your decision or even notice what you did.

Now if you were to fork twitter or any social app, you get the same benefit as you get with google docs, only that here you care about the people - because twitter is useless if you are the only person using it.

**but**, like you say:
> The people who admin a community are making moderation and ranking decisions which affect those indexes

So even if I can still read the data from the any user feed, the same as you might be able to read data from the twitter api,
the twitter user will not see your extra data, extra users or extra features or data in your new "community feed", so conversations wont happen. It is broken.

In some sense that has a similar implication (also a bit weaker) than a fork of a blockchain. Your forked TOKEN would not be listed and nobody would buy it from you if you spawn a bitcoin network over night and users will also not mine on your chain.

**Comment by pfrazee** [comment]
Yeah that's a fair point. This is one of the reasons that I have tied data-ID namespaces with the schemas -- because of the consensus problems of interactions. Despite the remaining messiness though, I still suspect forking will have _some_ utility, it's just not, you know, totally magic.

That said, do you have any suggestions to improving on it?  

> In some sense that has a similar implication (also a bit weaker) than a fork of a blockchain. Your forked TOKEN would not be listed and nobody would buy it from you if you spawn a bitcoin network over night and users will also not mine on your chain.

I still think that's pretty similar and not sure it's weaker, per se. Forks are always about social consensus to adopt the fork, and the only reason none of the coin forks have been successful is that none of them have really gotten social consensus.

It's true that code forks for "single player" apps are a little less complex but as soon as they're "multiplayer" they have the exact same problems.

**Comment by serapath** [comment]
Yes, maybe it is not even "weaker", which makes the problem worse.
I don't have any suggestion per se.

If it is possible to "audit" the server to detect fraud if you happen to run a "full node" (=replicating the servers community hypercore(s) ..nodevms style) and recompute the state of the database by re-executing all operations - is that how you audit?

...**then:** - in case of fraud - it would be nice to be able to reach out to all users and proof it to them together with maybe a proposal for a fork.

Maybe people would auto-follow the first proposal which broadcasts it.
If the person who detected fraud just to make up a new network with the intention to be fraudulent too - it will happen to them again, until somebody broadcasts a proposal for a fork that is not malicious anymore.

**Comment by serapath** [comment]
This does not need technical support in small communities, but you mentioned aiming for large communities (e.g. potentially twitter or facebook large) ...

**So it sounds like what it needs is a "leader election" process?**

ideally that process (even if expensive) will work well and everyone can understand why that is so, so it takes away the incentive from current leaders to even try to be malicious - which makes the expensiveness occur rarely, because they know it would just cost them their position and thats it.

**Comment by pfrazee** [comment]
The auditability idea is exactly what the constitutional networks section is about in the readme!

**Comment by serapath** [comment]
alright. thx. i'll check it out. guess i missed that :-)
